stellargraph/layer/node2vec.py

```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+node2vec
+
+"""
+__all__ = ["Node2Vec"]
+
+from tensorflow.keras import Input
+from tensorflow.keras.layers import Reshape, Embedding
+import math
+from tensorflow import keras
+import warnings
+from .misc import deprecated_model_function
+from ..mapper import Node2VecLinkGenerator, Node2VecNodeGenerator
+
+
+class Node2Vec:
+    """
+    Implementation of the node2vec algorithm of A. Grover and J. Leskovec with Keras layers.
+    see: https://snap.stanford.edu/node2vec/
+
+    The model minimally requires specification of the embedding size and a generator object.
+
+    Args:
+        emb_size (int): The dimension of node embeddings.
+        generator (Sequence): A NodeSequence or LinkSequence.
+        **kwargs: Parameters node_num(int) and multiplicity(int).
+    """
+
+    def __init__(self, emb_size, generator=None, **kwargs):
+
+        # Get the node_num from the generator if it is given
+        self.generator = generator
+        if generator is not None:
+            self._get_sizes_from_generator(generator)
+        else:
+            self._get_sizes_from_keywords(kwargs)
+
+        # Model parameters
+        self.emb_size = emb_size
+
+        # Initialise the embedding layers: input-to-hidden and hidden-to-output
+        input_initializer = keras.initializers.RandomUniform(minval=-1.0, maxval=1.0)
+        self.input_embedding = Embedding(
```

Comment:
If I understood correctly, you're referring to the "target node" as the `input` and the "context node" as `output`... Is this terminology standard for Node2Vec? 

I find it a bit confusing in the context of a keras model, since "input" and "output" already have a meaning (i.e. "input" tensors to the model vs the "output" tensors). Could we just name them as "target" and "context", similar to the way it's named in the generator?

Reply:
Also, I would've thought we actually want to only train a single embedding layer that's shared for both the target and context nodes? Otherwise we're saying that the embedding functions for the target node vs context node could be different.

Reply:
In the original word2vec implementation, two sets of embeddings i.e., `input_embedding` and `output_embedding` are learned for `target node` and `context node` respectively. Here, the `input_embedding` is actually the weight matrix from the input layer to the hidden representation layer, and the `output_embedding` is the weight matrix from hidden representation layer to the output layer. To be consistent with the word2vec algorithm, we should learn two different sets of embeddings for `target node` and `context node` with the purpose of making nodes sharing similar context nodes represented closely.

Reply:
Hmm okay, thanks for clarifying that. Do you think it's important that we use the terms `input_embedding` and `output_embedding`? Could it just be `target_embedding` and `context_embedding` instead? I'm just a bit concerned that it's confusing with the model's actual input and output tensors (i.e. `in_out_tensors`) - what do you think?

Reply:
That's a good idea! I will change the names to avoid confusing users.